Applications/spire/source/charting/cached_chart_model.cpp

```diff
@@ -36,90 +37,135 @@ QtPromise<std::vector<Candlestick>> CachedChartModel::load(ChartValue first,
       });
     }
   }
-  return load_data({first, last});
+  return load_data(get_gaps(first, last), first, last);
 }
 
 connection CachedChartModel::connect_candlestick_slot(
     const CandlestickSignal::slot_type& slot) const {
   return m_chart_model->connect_candlestick_slot(slot);
 }
 
+std::vector<ChartRange> CachedChartModel::get_gaps(ChartValue first,
+    ChartValue last) {
+  if(m_ranges.empty()) {
+    return {{ChartRange{first, last}}};
+  }
+  auto range_points = std::vector<ChartValue>({first, last});
+  for(auto& range : m_ranges) {
+    if(range.m_end < first) {
+      continue;
+    }
+    if(range.m_start >= last) {
+      break;
+    }
+    if(range.m_start <= first && range.m_end >= first) {
+      range_points.front() = range.m_end;
+    } else if(range.m_start <= last && range.m_end >= last) {
+      range_points.back() = range.m_start;
+    }
+  }
+  auto gaps = std::vector<ChartRange>();
+  for(auto i = std::size_t(0); i < range_points.size(); i += 2) {
+    gaps.push_back({range_points[i], range_points[i + 1]});
+  }
+  return gaps;
+}
+
+void CachedChartModel::insert_data(const std::vector<Candlestick>& data) {
+  if(data.empty()) {
+    return;
+  }
+  auto first = std::lower_bound(m_loaded_data.begin(),
+    m_loaded_data.end(), data.front().GetStart(), [] (const auto& index,
+        const auto& value) {
+      return index.GetStart() < value;
+    });
+  auto last = std::lower_bound(m_loaded_data.begin(),
```

Comment:
Which of the following ranges do you want to erase?

[first, last]
(first, last)
[first, last)
(first, last]

Reply:
I want to erase [first, last].

The assignment of 'last' gets the element past the element I'm looking for. So in when I call erase I'm basically saying erase [first, last + 1).